File: ios/LookSeeProto/Services/OfflineMediaManager.swift
//
//  OfflineMediaManager.swift
//  LookSeeProto
//
//  Created by Angel Pineda on 6/18/26.
//  Updated for Outbox/Queue functionality
//


import Foundation
import CoreLocation
import SwiftUI
import Combine
import AVFoundation

@MainActor
class OfflineMediaManager: ObservableObject {
    static let shared = OfflineMediaManager()
    
    @Published var archivedItems: [ArchivedMedia] = []
    
    // In-memory cache to keep negative photos alive during your app session
    @Published var negativeCache: [UUID: [CapturedNegativePhoto]] = [:]
    
    private let ledgerKey = "LookSeeArchiveLedger"
    
    private init() {
        loadArchive()
    }
    
    func getDocumentsDirectory() -> URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }
    
    func getFileURL(for media: ArchivedMedia) -> URL {
        getDocumentsDirectory().appendingPathComponent(media.fileName)
    }
    
    func getThumbnailURL(for media: ArchivedMedia) -> URL {
        getDocumentsDirectory().appendingPathComponent(media.thumbnailFileName)
    }
    
    func getNegativeVideoURL(for media: ArchivedMedia) -> URL? {
        guard let negName = media.negativeVideoFileName else { return nil }
        return getDocumentsDirectory().appendingPathComponent(negName)
    }
    
    // 🚀 THE FIX: Bumps an item to the absolute front of the queue by spoofing an old date
    func prioritizeAndRetry(media: ArchivedMedia) {
        if let index = archivedItems.firstIndex(where: { $0.id == media.id }) {
            // Date(timeIntervalSince1970: 0) makes it the oldest possible item, forcing it to the front!
            archivedItems[index].dateSaved = Date(timeIntervalSince1970: 0)
            archivedItems.sort { $0.dateSaved < $1.dateSaved }
            saveArchive()
        }
    }
    
    // MARK: - Archive Video (Queue) - Background Optimized
    func archiveVideo(
        tempURL: URL,
        lat: Double,
        lon: Double,
        landmarkId: String?,
        label: String,
        shortDesc: String,
        userDesc: String?,
        negativeVideoURL: URL?,
        isTier2: Bool = false
    ) async -> ArchivedMedia? {
        
        let uniqueID = UUID().uuidString
        let fileName = uniqueID + ".mov"
        let thumbName = uniqueID + "_thumb.jpg"
        
        // Grab the directory on the main thread to pass to the background
        let docsDir = getDocumentsDirectory()
        let permanentURL = docsDir.appendingPathComponent(fileName)
        
        // Detach heavy file I/O to a background thread to prevent UI lag
        let newEntry = await Task.detached(priority: .userInitiated) { () -> ArchivedMedia? in
            var permanentNegName: String? = nil
            
            do {
                // 1. Copy Positive Video
                try FileManager.default.copyItem(at: tempURL, to: permanentURL)
                
                // 2. Generate Thumbnail (Heavy CPU Task)
                let asset = AVAsset(url: permanentURL)
                let generator = AVAssetImageGenerator(asset: asset)
                generator.appliesPreferredTrackTransform = true
                
                if let cgImage = try? generator.copyCGImage(at: .zero, actualTime: nil) {
                    let thumbnail = UIImage(cgImage: cgImage)
                    if let data = thumbnail.jpegData(compressionQuality: 0.8) {
                        let thumbURL = docsDir.appendingPathComponent(thumbName)
                        try? data.write(to: thumbURL)
                    }
                }
                
                // 3. Copy Negative Video
                if let negURL = negativeVideoURL {
                    let negName = uniqueID + "_negative.mov"
                    let permanentNegURL = docsDir.appendingPathComponent(negName)
                    try FileManager.default.copyItem(at: negURL, to: permanentNegURL)
                    permanentNegName = negName
                }
                
                // 4. Create Queue Record
                return ArchivedMedia(
                    title: label,
                    fileName: fileName,
                    thumbnailFileName: thumbName,
                    isVideo: true,
                    latitude: lat,
                    longitude: lon,
                    dateSaved: Date(),
                    isFavorite: false,
                    landmarkId: landmarkId,
                    savedLabel: label,
                    savedDescription: shortDesc,
                    savedUserDescription: userDesc,
                    negativeVideoFileName: permanentNegName,
                    isTier2: isTier2
                )
                
            } catch {
                print("❌ Failed to archive video: \(error)")
                return nil
            }
        }.value
        
        // Back on Main Thread: Update UI immediately
        if let entry = newEntry {
            self.archivedItems.append(entry)
            self.archivedItems.sort { $0.dateSaved < $1.dateSaved } // 🚀 Keep chronological
            self.saveArchive()
        }
        
        return newEntry
    }
    
    // MARK: - Archive Photo (Queue) - Background Optimized
    func archivePhoto(
        image: UIImage,
        lat: Double,
        lon: Double,
        landmarkId: String?,
        label: String,
        shortDesc: String,
        userDesc: String?,
        negativeVideoURL: URL?,
        isTier2: Bool = false
    ) async -> ArchivedMedia? {
        
        let uniqueID = UUID().uuidString
        let fileName = uniqueID + ".jpg"
        let thumbName = uniqueID + "_thumb.jpg"
        let docsDir = getDocumentsDirectory()
        
        // Detach heavy image compression to background thread
        let newEntry = await Task.detached(priority: .userInitiated) { () -> ArchivedMedia? in
            var permanentNegName: String? = nil
            
            guard let data = image.jpegData(compressionQuality: 0.8),
                  let thumbData = image.jpegData(compressionQuality: 0.3) else {
                return nil
            }
            
            do {
                let url = docsDir.appendingPathComponent(fileName)
                let thumbURL = docsDir.appendingPathComponent(thumbName)
                
                try data.write(to: url)
                try thumbData.write(to: thumbURL)
                
                // Copy Negative Video
                if let negURL = negativeVideoURL {
                    let negName = uniqueID + "_negative.mov"
                    let permanentNegURL = docsDir.appendingPathComponent(negName)
                    try FileManager.default.copyItem(at: negURL, to: permanentNegURL)
                    permanentNegName = negName
                }
                
                return ArchivedMedia(
                    title: label,
                    fileName: fileName,
                    thumbnailFileName: thumbName,
                    isVideo: false,
                    latitude: lat,
                    longitude: lon,
                    dateSaved: Date(),
                    isFavorite: false,
                    landmarkId: landmarkId,
                    savedLabel: label,
                    savedDescription: shortDesc,
                    savedUserDescription: userDesc,
                    negativeVideoFileName: permanentNegName,
                    isTier2: isTier2
                )
            } catch {
                print("❌ Failed to archive photo: \(error)")
                return nil
            }
        }.value
        
        if let entry = newEntry {
            self.archivedItems.append(entry)
            self.archivedItems.sort { $0.dateSaved < $1.dateSaved } // 🚀 Keep chronological
            self.saveArchive()
        }
        
        return newEntry
    }
    
    // MARK: - Draft Updates
    func updateDraft(media: ArchivedMedia, label: String, shortDesc: String, userDesc: String?) {
        if let index = archivedItems.firstIndex(where: { $0.id == media.id }) {
            archivedItems[index].savedLabel = label
            archivedItems[index].savedDescription = shortDesc
            archivedItems[index].savedUserDescription = userDesc
            archivedItems[index].title = label
            saveArchive()
        }
    }
    
    func renameArchive(media: ArchivedMedia, newTitle: String) {
        if let index = archivedItems.firstIndex(where: { $0.id == media.id }) {
            archivedItems[index].title = newTitle
            saveArchive()
        }
    }
    
    func toggleFavorite(media: ArchivedMedia) {
        if let index = archivedItems.firstIndex(where: { $0.id == media.id }) {
            let current = archivedItems[index].isFavorite ?? false
            archivedItems[index].isFavorite = !current
            saveArchive()
        }
    }
    
    // MARK: - Delete (Background Optimized)
    func deleteArchive(media: ArchivedMedia) {
        let fileURL = getFileURL(for: media)
        let thumbURL = getThumbnailURL(for: media)
        let negURL = getNegativeVideoURL(for: media)
        
        // Tell the hard drive to delete the heavy files in the background
        Task.detached(priority: .background) {
            try? FileManager.default.removeItem(at: fileURL)
            try? FileManager.default.removeItem(at: thumbURL)
            if let nURL = negURL {
                try? FileManager.default.removeItem(at: nURL)
            }
        }
        
        // Remove from the UI immediately so it feels instant
        negativeCache.removeValue(forKey: media.id)
        archivedItems.removeAll { $0.id == media.id }
        saveArchive()
    }
    
    private func saveArchive() {
        if let encoded = try? JSONEncoder().encode(archivedItems) {
            UserDefaults.standard.set(encoded, forKey: ledgerKey)
        }
    }
    
    private func loadArchive() {
        if let savedData = UserDefaults.standard.data(forKey: ledgerKey),
           let decoded = try? JSONDecoder().decode([ArchivedMedia].self, from: savedData) {
            // 🚀 THE FIX: Ensure it is sorted oldest-first on load
            archivedItems = decoded.sorted { $0.dateSaved < $1.dateSaved }
        }
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/services/OfflineMediaManager.kt
package looksee.angelll.com.services

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.net.Uri
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

object OfflineMediaManager {

    private const val PREFS_NAME = "LookSeeOfflinePrefs"
    private const val LEDGER_KEY = "LookSeeArchiveLedger"

    // 🚀 Reactive States for Compose UI
    private val _archivedItems = MutableStateFlow<List<ArchivedMedia>>(emptyList())
    val archivedItems: StateFlow<List<ArchivedMedia>> = _archivedItems.asStateFlow()

    // In-memory cache to keep negative photos alive during your app session
    private val _negativeCache = MutableStateFlow<Map<String, List<CapturedNegativePhoto>>>(emptyMap())
    val negativeCache: StateFlow<Map<String, List<CapturedNegativePhoto>>> = _negativeCache.asStateFlow()

    // Call this once from your MainActivity or Application class: OfflineMediaManager.initialize(context)
    fun initialize(context: Context) {
        loadArchive(context)
    }

    private fun getDocumentsDirectory(context: Context): File {
        return context.filesDir
    }

    fun getFile(context: Context, media: ArchivedMedia): File {
        return File(getDocumentsDirectory(context), media.fileName)
    }

    fun getThumbnail(context: Context, media: ArchivedMedia): File {
        return File(getDocumentsDirectory(context), media.thumbnailFileName)
    }

    fun getNegativeVideo(context: Context, media: ArchivedMedia): File? {
        val negName = media.negativeVideoFileName ?: return null
        return File(getDocumentsDirectory(context), negName)
    }

    // 🚀 THE FIX: Bumps an item to the absolute front of the queue by spoofing an old date
    fun prioritizeAndRetry(context: Context, media: ArchivedMedia) {
        _archivedItems.update { currentList ->
            val mutableList = currentList.toMutableList()
            val index = mutableList.indexOfFirst { it.id == media.id }
            if (index != -1) {
                // 0L makes it the oldest possible item (1970), forcing it to the front!
                val updatedItem = mutableList[index].copy(dateSaved = 0L)
                mutableList[index] = updatedItem
                mutableList.sortBy { it.dateSaved }
            }
            mutableList
        }
        saveArchive(context)
    }

    // MARK: - Archive Video (Queue) - Background Optimized
    suspend fun archiveVideo(
        context: Context,
        tempUri: Uri,
        lat: Double,
        lon: Double,
        landmarkId: String?,
        label: String,
        shortDesc: String,
        userDesc: String?,
        negativeVideoUri: Uri?,
        isTier2: Boolean = false
    ): ArchivedMedia? = withContext(Dispatchers.IO) {
        val uniqueID = UUID.randomUUID().toString()
        val fileName = "$uniqueID.mov"
        val thumbName = "${uniqueID}_thumb.jpg"

        val docsDir = getDocumentsDirectory(context)
        val permanentFile = File(docsDir, fileName)
        var permanentNegName: String? = null

        try {
            // 1. Copy Positive Video
            context.contentResolver.openInputStream(tempUri)?.use { input ->
                permanentFile.outputStream().use { output ->
                    input.copyTo(output)
                }
            }

            // 2. Generate Thumbnail (Heavy CPU Task)
            val retriever = MediaMetadataRetriever()
            try {
                retriever.setDataSource(context, tempUri)
                val bitmap = retriever.getFrameAtTime(0, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
                if (bitmap != null) {
                    val thumbFile = File(docsDir, thumbName)
                    thumbFile.outputStream().use { out ->
                        bitmap.compress(Bitmap.CompressFormat.JPEG, 80, out)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                retriever.release()
            }

            // 3. Copy Negative Video
            if (negativeVideoUri != null) {
                val negName = "${uniqueID}_negative.mov"
                val permanentNegFile = File(docsDir, negName)
                context.contentResolver.openInputStream(negativeVideoUri)?.use { input ->
                    permanentNegFile.outputStream().use { output ->
                        input.copyTo(output)
                    }
                }
                permanentNegName = negName
            }

            // 4. Create Queue Record
            val newEntry = ArchivedMedia(
                id = uniqueID,
                title = label,
                fileName = fileName,
                thumbnailFileName = thumbName,
                isVideo = true,
                latitude = lat,
                longitude = lon,
                dateSaved = System.currentTimeMillis(),
                isFavorite = false,
                landmarkId = landmarkId,
                savedLabel = label,
                savedDescription = shortDesc,
                savedUserDescription = userDesc,
                negativeVideoFileName = permanentNegName,
                isTier2 = isTier2
            )

            // Update state flow on main thread equivalent
            _archivedItems.update { current ->
                val newList = current.toMutableList()
                newList.add(newEntry)
                newList.sortBy { it.dateSaved } // 🚀 Keep chronological
                newList
            }
            saveArchive(context)

            return@withContext newEntry

        } catch (e: Exception) {
            println("❌ Failed to archive video: ${e.localizedMessage}")
            return@withContext null
        }
    }

    // MARK: - Archive Photo (Queue) - Background Optimized
    suspend fun archivePhoto(
        context: Context,
        image: Bitmap,
        lat: Double,
        lon: Double,
        landmarkId: String?,
        label: String,
        shortDesc: String,
        userDesc: String?,
        negativeVideoUri: Uri?,
        isTier2: Boolean = false
    ): ArchivedMedia? = withContext(Dispatchers.IO) {
        val uniqueID = UUID.randomUUID().toString()
        val fileName = "$uniqueID.jpg"
        val thumbName = "${uniqueID}_thumb.jpg"
        val docsDir = getDocumentsDirectory(context)
        var permanentNegName: String? = null

        try {
            val permanentFile = File(docsDir, fileName)
            val thumbFile = File(docsDir, thumbName)

            // Write full image
            permanentFile.outputStream().use { out ->
                image.compress(Bitmap.CompressFormat.JPEG, 80, out)
            }

            // Generate and write thumbnail
            val ratio = Math.min(400f / image.width, 400f / image.height)
            val width = Math.round(ratio * image.width)
            val height = Math.round(ratio * image.height)
            val thumbBitmap = Bitmap.createScaledBitmap(image, width, height, true)

            thumbFile.outputStream().use { out ->
                thumbBitmap.compress(Bitmap.CompressFormat.JPEG, 30, out)
            }

            // Copy Negative Video
            if (negativeVideoUri != null) {
                val negName = "${uniqueID}_negative.mov"
                val permanentNegFile = File(docsDir, negName)
                context.contentResolver.openInputStream(negativeVideoUri)?.use { input ->
                    permanentNegFile.outputStream().use { output ->
                        input.copyTo(output)
                    }
                }
                permanentNegName = negName
            }

            val newEntry = ArchivedMedia(
                id = uniqueID,
                title = label,
                fileName = fileName,
                thumbnailFileName = thumbName,
                isVideo = false,
                latitude = lat,
                longitude = lon,
                dateSaved = System.currentTimeMillis(),
                isFavorite = false,
                landmarkId = landmarkId,
                savedLabel = label,
                savedDescription = shortDesc,
                savedUserDescription = userDesc,
                negativeVideoFileName = permanentNegName,
                isTier2 = isTier2
            )

            _archivedItems.update { current ->
                val newList = current.toMutableList()
                newList.add(newEntry)
                newList.sortBy { it.dateSaved } // 🚀 Keep chronological
                newList
            }
            saveArchive(context)

            return@withContext newEntry

        } catch (e: Exception) {
            println("❌ Failed to archive photo: ${e.localizedMessage}")
            return@withContext null
        }
    }

    // MARK: - Draft Updates
    fun updateDraft(context: Context, media: ArchivedMedia, label: String, shortDesc: String, userDesc: String?) {
        _archivedItems.update { currentList ->
            currentList.map { item ->
                if (item.id == media.id) {
                    item.copy(
                        savedLabel = label,
                        savedDescription = shortDesc,
                        savedUserDescription = userDesc,
                        title = label
                    )
                } else item
            }
        }
        saveArchive(context)
    }

    fun renameArchive(context: Context, media: ArchivedMedia, newTitle: String) {
        _archivedItems.update { currentList ->
            currentList.map { item ->
                if (item.id == media.id) item.copy(title = newTitle) else item
            }
        }
        saveArchive(context)
    }

    fun toggleFavorite(context: Context, media: ArchivedMedia) {
        _archivedItems.update { currentList ->
            currentList.map { item ->
                if (item.id == media.id) item.copy(isFavorite = !item.isFavorite) else item
            }
        }
        saveArchive(context)
    }

    // MARK: - Delete (Background Optimized)
    fun deleteArchive(context: Context, media: ArchivedMedia) {
        val file = getFile(context, media)
        val thumb = getThumbnail(context, media)
        val negFile = getNegativeVideo(context, media)

        // Tell the hard drive to delete the heavy files in the background
        CoroutineScope(Dispatchers.IO).launch {
            file.delete()
            thumb.delete()
            negFile?.delete()
        }

        // Remove from the UI immediately so it feels instant
        _negativeCache.update { currentMap ->
            val newMap = currentMap.toMutableMap()
            newMap.remove(media.id)
            newMap
        }

        _archivedItems.update { currentList ->
            currentList.filter { it.id != media.id }
        }

        saveArchive(context)
    }

    // MARK: - Native JSON Persistence
    private fun saveArchive(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val jsonArray = JSONArray()

        _archivedItems.value.forEach { media ->
            val obj = JSONObject().apply {
                put("id", media.id)
                put("title", media.title)
                put("fileName", media.fileName)
                put("thumbnailFileName", media.thumbnailFileName)
                put("isVideo", media.isVideo)
                put("latitude", media.latitude)
                put("longitude", media.longitude)
                put("dateSaved", media.dateSaved)
                put("isFavorite", media.isFavorite)
                put("landmarkId", media.landmarkId ?: JSONObject.NULL)
                put("savedLabel", media.savedLabel)
                put("savedDescription", media.savedDescription)
                put("savedUserDescription", media.savedUserDescription ?: JSONObject.NULL)
                put("negativeVideoFileName", media.negativeVideoFileName ?: JSONObject.NULL)
                put("isTier2", media.isTier2)
            }
            jsonArray.put(obj)
        }

        prefs.edit().putString(LEDGER_KEY, jsonArray.toString()).apply()
    }

    private fun loadArchive(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val jsonString = prefs.getString(LEDGER_KEY, null) ?: return

        try {
            val jsonArray = JSONArray(jsonString)
            val loadedItems = mutableListOf<ArchivedMedia>()

            for (i in 0 until jsonArray.length()) {
                val obj = jsonArray.getJSONObject(i)
                loadedItems.add(
                    ArchivedMedia(
                        id = obj.getString("id"),
                        title = obj.getString("title"),
                        fileName = obj.getString("fileName"),
                        thumbnailFileName = obj.getString("thumbnailFileName"),
                        isVideo = obj.getBoolean("isVideo"),
                        latitude = obj.getDouble("latitude"),
                        longitude = obj.getDouble("longitude"),
                        dateSaved = obj.getLong("dateSaved"),
                        isFavorite = obj.optBoolean("isFavorite", false),
                        landmarkId = if (obj.isNull("landmarkId")) null else obj.getString("landmarkId"),
                        savedLabel = obj.getString("savedLabel"),
                        savedDescription = obj.getString("savedDescription"),
                        savedUserDescription = if (obj.isNull("savedUserDescription")) null else obj.getString("savedUserDescription"),
                        negativeVideoFileName = if (obj.isNull("negativeVideoFileName")) null else obj.getString("negativeVideoFileName"),
                        isTier2 = obj.optBoolean("isTier2", false)
                    )
                )
            }

            // 🚀 THE FIX: Ensure it is sorted oldest-first on load
            _archivedItems.value = loadedItems.sortedBy { it.dateSaved }

        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}

// MARK: - Stubs (Ensure these match your actual data models)
data class ArchivedMedia(
    val id: String,
    val title: String,
    val fileName: String,
    val thumbnailFileName: String,
    val isVideo: Boolean,
    val latitude: Double,
    val longitude: Double,
    val dateSaved: Long,
    val isFavorite: Boolean,
    val landmarkId: String?,
    val savedLabel: String,
    val savedDescription: String,
    val savedUserDescription: String?,
    val negativeVideoFileName: String?,
    val isTier2: Boolean
)

data class CapturedNegativePhoto(
    val id: String = UUID.randomUUID().toString(),
    val uri: Uri
)